### PR TITLE
Encode redirect URL if needed on oauth callback request

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
Some SCM providers like BItBucket Server decode the callback url so that cause `IllegalArgumentException` error. Catch the error and decode the redirect url.

### What issues does this PR fix or reference?

https://issues.redhat.com/browse/CRW-5663

### How to test this PR?

1. Deploy che with the PR image: `quay.io/eclipse/che-server:pr-654`
2. Configure oauth for a BitBucket Server instance.
3. Create a workspace from the BitBucket Server repository, accept the oauth agreement.
4. Stop the workspace and revoke the agreement from the oauth application.
5. Start the workspace and accept the agreement again.

See: workspace starts successfully.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth/EmbeddedOAuthAPI.java [hunk 1]
```diff
@@ -123,20 +125,25 @@ public Response callback(UriInfo uriInfo, @Nullable List<String> errorValues)
       LOG.error(e.getMessage(), e);
     }
     final String redirectAfterLogin = getParameter(params, "redirect_after_login");
-    return Response.temporaryRedirect(URI.create(redirectAfterLogin)).build();
+    URI uri;
+    try {
+      uri = URI.create(redirectAfterLogin);
+    } catch (IllegalArgumentException e) {
+      // the redirectUrl was decoded by the CSM provider, so we need to encode it back.
```
Comment (tolusha) [inline wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth/EmbeddedOAuthAPI.java:132]: It is not clear why IllegalArgumentException is the indicator of decoded url

Comment (vinokurig) [reply wsmaster/che-core-api-auth/src/main/java/org/eclipse/che/security/oauth/EmbeddedOAuthAPI.java:132]: This exception is thrown when we try to create a new URL from a string with unsupported characters like `{` and `}`. Initially such symbols are encoded, but BitBucket Server decodes them so they become unsupported.

**Review by tolusha — APPROVED** [review]

**Comment by artaleks9** [comment]
Verified on Eclipse Che with `quay.io/eclipse/che-server:pr-654` - the functionality works as expected.

**Review by artaleks9 — APPROVED** [review]

**Pull request merged**